### PR TITLE
🌱 clusterctl/client/cert_manager: improve shouldUpgrade

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:

Make shouldUpgrade() accept the list of objects to be installed too.
Compare the installed and to-be-installed objects in case the
installed and to-be-installed versions are the same. If their length is
different, assume that an upgrade is required.

Update unit tests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/10389
(the linked issue was a but unclear what's going on, but we established that shouldUpgrade can check len of objects too)

/area clusterctl
/kind feature

### CODE REVIEW

**Comment by neolit123** [comment]
/cc @chrischdi @sbueringer 
PTAL

File: cmd/clusterctl/client/cluster/cert_manager_test.go [hunk 1]
```diff
@@ -398,6 +399,27 @@ func Test_shouldUpgrade(t *testing.T) {
 			want:            false,
 			wantErr:         false,
 		},
+		{
+			name: "Version is newer, should not upgrade, but objects to install are a different size",
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply cmd/clusterctl/client/cluster/cert_manager_test.go:403]: cc @fabriziopandini @chrischdi 

Not sure what we want to do in this case. Maybe we only want to consider the obj length check for desiredVersion >= current version?

File: cmd/clusterctl/client/cluster/cert_manager.go [hunk 2]
```diff
@@ -334,6 +353,12 @@ func (cm *certManagerClient) shouldUpgrade(objs []unstructured.Unstructured) (st
...
 
+	// If the number of avaiable objects and objects to install differ, return true, but first fetch the current version.
+	diffObjLen := false
+	if len(objs) != len(installObjs) {
+		diffObjLen = true
+	}
+
 	needUpgrade := false
 	currentVersion := ""
 	for i := range objs {
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/cert_manager.go:364]: Let's fix the comment in l.392 I think it's misleading (or at least very hard to understand)

File: cmd/clusterctl/client/cluster/cert_manager_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrischdi) [reply cmd/clusterctl/client/cluster/cert_manager_test.go:403]: +1 to not downgrade!

[the remark this answers is not in the document]
Comment (neolit123) [reply cmd/clusterctl/client/cluster/cert_manager_test.go:403]: updating the logic and unit tests to:
- upgrade if the installed and to-be-installed versions are equal but the installed vs to-beinstalled object counts are different. if the versions are not the same do not consider checking object counts.

**Comment by sbueringer** [comment]
Thx!

/lgtm

/assign @fabriziopandini

**Review by chrischdi — COMMENTED** [review]
/approve

Thanks for working on this!

File: cmd/clusterctl/client/cluster/cert_manager.go [hunk 3]
```diff
@@ -269,27 +288,16 @@ func (cm *certManagerClient) EnsureLatestVersion(ctx context.Context) error {
...
-	}
-
+func (cm *certManagerClient) migrateCRDs(ctx context.Context, installObj []unstructured.Unstructured) error {
 	c, err := cm.proxy.NewClient(ctx)
 	if err != nil {
 		return err
 	}
 
-	return NewCRDMigrator(c).Run(ctx, objs)
+	return NewCRDMigrator(c).Run(ctx, installObj)
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/cert_manager.go:300]: q: why are passing installed objects here?
If I remember well CRD migrator expect to get in input the objects that are going to be installed to determine what is going to change

Comment (chrischdi) [reply cmd/clusterctl/client/cluster/cert_manager.go:300]: installObj are the objects to-be installed, not the one which exist :-)

It's the same as before, the output of: `cm.getManifestObjs(ctx, config)`

**Pull request merged**